### PR TITLE
Sql unit tests options

### PR DESCRIPTION
**Pull request by tim-harper**

### CODE REVIEW

File: internal/datastore/datastore.go [hunk 1]
```diff
@@ -168,53 +159,18 @@ func (d *delphisDB) initializeStatements(ctx context.Context) (err error) {
...
-		) VALUES ($1, $2, $3, $4, $5)
-		RETURNING
-			id,
-			created_at,
-			updated_at,
-			discussion_id,
-			participant_id,
-			post_content_id,
-			quoted_post_id;`); err != nil {
+	if d.prepStmts.putPostStmt, err = d.pg.PrepareContext(ctx, putPostString); err != nil {
```
Comment (tim-harper) [inline internal/datastore/datastore.go:162]: moved the prepare statements into their own file. this greatly helps with sqlmock as we can just pass in the exact string. will also help keep our tests and queries in sync

File: internal/datastore/datastore_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package datastore
+
```
Comment (tim-harper) [inline internal/datastore/datastore_test.go:2]: this file is for creating the testing database

File: internal/datastore/post.go [hunk 3]
```diff
@@ -21,7 +21,7 @@ func (d *delphisDB) PutPost(ctx context.Context, post model.Post) (*model.Post,
 		post.ID,
 		post.DiscussionID,
 		post.ParticipantID,
-		post.PostContentID,
+		post.PostContent.ID,
```
Comment (tim-harper) [inline internal/datastore/post.go:24]: can clean up the IDs since we aren't using gorm here

File: internal/datastore/post_db_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
+package datastore
+
```
Comment (tim-harper) [inline internal/datastore/post_db_test.go:2]: post_db_test uses the separate db to test POST

File: internal/datastore/post_test.go [hunk 5]
```diff
@@ -1,59 +1,68 @@
 package datastore
 
 import (
+	"context"
+	"regexp"
 	"testing"
+	"time"
+
+	"github.com/DATA-DOG/go-sqlmock"
+	"github.com/jinzhu/gorm"
```
Comment (tim-harper) [inline internal/datastore/post_test.go:10]: this file uses sqlmock

File: internal/datastore/test_tables.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	"github.com/sirupsen/logrus"
+)
+
+type TestData struct {
+	Discussions  []model.Discussion
+	Participants []model.Participant
+	Posts        []model.Post
+}
+
+func (d *delphisDB) CreateTestTables(ctx context.Context, data TestData) (func() error, error) {
```
Comment (tim-harper) [inline internal/datastore/test_tables.go:17]: this file builds and populates the testing db

File: internal/datastore/tests/models.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
+package tests
+
+import (
+	"github.com/nedrocks/delphisbe/graph/model"
+)
+
```
Comment (tim-harper) [inline internal/datastore/tests/models.go:6]: handy file for testing data

your `alia uuid` script came in very clutch

File: internal/datastore/post_db_test.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	// Initialize the test DB once per file
+	db, close, err := MakeDatastore(context.Background(), data)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer close()
+
+	// Loop over table tests
+	tests := []struct {
```
Comment (nedrocks) [inline internal/datastore/post_db_test.go:45]: I know I have one or two tests structured with this format but I find it slightly harder to read than convey. I'm not wedded to one side or the other, but it's harder to convey (harhar) what you're actually testing as much as with Convey.

File: internal/datastore/post_test.go [hunk 9]
```diff
@@ -1,59 +1,71 @@
...
+			encoder:   nil,
+		}
+		defer db.Close()
+
+		// Prepare statements
+		mock.ExpectPrepare(regexp.QuoteMeta(putPostString))
+		mock.ExpectPrepare(regexp.QuoteMeta(getPostsByDiscussionIDString))
+		mock.ExpectPrepare(regexp.QuoteMeta(putPostContentsString))
+
+		Convey("when the works", func() {
```
Comment (nedrocks) [inline internal/datastore/post_test.go:57]: lol at the name of this test

File: internal/datastore/datastore_test.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func createTestDatabase(ctx context.Context, url string) (string, error) {
+	db, err := gorm.Open("postgres", url)
+	if err != nil {
+		return "", err
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	name := "tests_" + strconv.Itoa(rand.Int())
```
Comment (nedrocks) [inline internal/datastore/datastore_test.go:74]: I like this!

File: internal/datastore/test_tables.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	return d.close, nil
+}
+
+func (d *delphisDB) createTestTables(ctx context.Context) error {
+	d.readyMu.Lock()
+	defer d.readyMu.Unlock()
+
+	createTableQueries := []string{
```
Comment (nedrocks) [inline internal/datastore/test_tables.go:35]: Maybe I'm ignorant here but why not use the migrations in _scripts/migrations? It feels like this way the test db will diverge from the true DB schema over time.

**Pull request merged**